### PR TITLE
`eth_getBlockReceipts` should accept block hashes

### PR DESCRIPTION
**Pull request by naps62**
## Motivation

While most documentation for `eth_getBlockReceipts` suggests the argument should be a block number, the reality seems very different: a lot of nodes support block hashes as well, and subgraph even depends on this behaviour to work. [The official RPC specification](https://ethereum.github.io/execution-apis/api-documentation/) seems to indicate that it should be supported

Quick test, using quicknode's API:

```
curl https://docs-demo.quiknode.pro/ \
  -X POST \
  -H "Content-Type: application/json" \
  --data '{"method":"eth_getBlockReceipts","params":["0x9732177306812f1c165bec8f1ee957a5fc4fea4498b82d842aee4f7c28fb9d08"],"id":1,"jsonrpc":"2.0"}'
```

Anvil currently only accepts a block number. This means that currently testing a subgraph with a local anvil instance does not work. Which is the exact use case that led me here

## Solution

Change `eth_getBlockReceipts` to receive a `BlockId` instead, which supports both behaviors.
This was tested locally with a private subgraph we're currently building, and this branch is currently being used successfully to run it locally

### CODE REVIEW

**Comment by zerosnacks** [comment]
Hi @naps62 thanks for your PR!

For comparison, it looks like we are using `BlockNumberOrTag` in https://github.com/alloy-rs/alloy/blob/c3ccf7e2c5f9720fa6c193a9ea918aabf88aa6a2/crates/provider/src/provider/trait.rs#L323-L329

According to the documentation you linked `BlockId` indeed seems to be preferred here.

**Comment by zerosnacks** [comment]
FYI: created a ticket here: https://github.com/alloy-rs/alloy/issues/1134 to track evaluating of the correctness of `BlockId` vs `BlockNumberOrTag` in Alloy

File: crates/anvil/src/eth/backend/mem/mod.rs [hunk 1]
```diff
@@ -1758,6 +1758,7 @@ impl Backend {
                 },
             };
 
+        dbg!(current, requested);
```
Comment (zerosnacks) [inline crates/anvil/src/eth/backend/mem/mod.rs:1761]:
```suggestion
```

File: crates/anvil/src/eth/api.rs [hunk 2]
```diff
@@ -262,7 +262,7 @@ impl EthApi {
                 self.transaction_receipt(tx).await.to_rpc_result()
             }
             EthRequest::EthGetBlockReceipts(number) => {
-                self.block_receipts(number).await.to_rpc_result()
+                self.block_receipts(number.into()).await.to_rpc_result()
```
Comment (zerosnacks) [inline crates/anvil/src/eth/api.rs:265]:
```suggestion
                self.block_receipts(number).await.to_rpc_result()
```

File: crates/anvil/tests/it/fork.rs [hunk 3]
```diff
@@ -1015,12 +1015,19 @@ async fn test_block_receipts() {
...
     assert!(receipts.is_some());
 
     // Receipts from a block in the future (14608401)
-    let receipts = api.block_receipts(BlockNumberOrTag::Number(BLOCK_NUMBER + 1)).await.unwrap();
+    let receipts =
+        api.block_receipts(BlockNumberOrTag::Number(BLOCK_NUMBER + 1).into()).await.unwrap();
     assert!(receipts.is_none());
+
+    // Receipts from a block hash (14608400)
+    let hash = B256::from_str("0x4c1c76f89cfe4eb503b09a0993346dd82865cac9d76034efc37d878c66453f0a")
```
Comment (zerosnacks) [inline crates/anvil/tests/it/fork.rs:1027]: you can use the `b256!` macro here

File: crates/anvil/tests/it/fork.rs [hunk 4]
```diff
@@ -19,7 +19,7 @@ use foundry_common::provider::get_http_provider;
 use foundry_config::Config;
 use foundry_test_utils::rpc::{self, next_http_rpc_endpoint};
 use futures::StreamExt;
-use std::{sync::Arc, thread::sleep, time::Duration};
+use std::{str::FromStr as _, sync::Arc, thread::sleep, time::Duration};
```
Comment (zerosnacks) [inline crates/anvil/tests/it/fork.rs:22]: Likely redundant import

**Review by mattsse — COMMENTED** [review]
BlockId is actually correct

image

so we should update in alloy as well

File: crates/anvil/tests/it/fork.rs (hunk 3 above)
Comment (naps62) [reply crates/anvil/tests/it/fork.rs:1027]: wasn't aware. thanks!

File: crates/anvil/tests/it/fork.rs (hunk 4 above)
Comment (naps62) [reply crates/anvil/tests/it/fork.rs:22]: indeed

**Review by mattsse — APPROVED** [review]
lgtm, pending @zerosnacks

**Review by zerosnacks — APPROVED** [review]
LGTM! Thanks for flagging and the fix @naps62

**Pull request merged**